File: app/src/main/java/com/vitaliyhtc/dagger2investigation/data/model/mapper/ProductsMapper.java
package com.vitaliyhtc.dagger2investigation.data.model.mapper;

import com.vitaliyhtc.dagger2investigation.domain.model.Product;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.functions.Function;

public class ProductsMapper implements Function<List<com.vitaliyhtc.dagger2investigation.data.model.response.Product>, List<Product>> {

    private ProductMapper mPm = new ProductMapper();

    @Override
    public List<Product> apply(List<com.vitaliyhtc.dagger2investigation.data.model.response.Product> products) throws Exception {
        List<Product> list = new ArrayList<>();
        for (com.vitaliyhtc.dagger2investigation.data.model.response.Product product1 : products) {
            list.add(mPm.apply(product1));
        }
        return list;
    }
}

File: app/src/main/java/com/vitaliyhtc/dagger2investigation/data/model/mapper/ProductsMapper.kt
package com.vitaliyhtc.dagger2investigation.data.model.mapper

import com.vitaliyhtc.dagger2investigation.domain.model.Product
import io.reactivex.functions.Function
import com.vitaliyhtc.dagger2investigation.data.model.response.Product as ProductApi

class ProductsMapper : Function<List<ProductApi>, List<Product>> {

    private val mapper = ProductMapper()

    override fun apply(products: List<ProductApi>): List<Product> {
        return products.map { mapper.apply(it) }
    }
}